src/structures/GuildTemplate.js

```diff
@@ -0,0 +1,197 @@
+'use strict';
+
+const Base = require('./Base');
+const { Events } = require('../util/Constants');
+const DataResolver = require('../util/DataResolver');
+
+/**
+ * Represents the template for a guild.
+ * @extends {Base}
+ */
+class GuildTemplate extends Base {
+  /**
+   * @param {Client} client The instantiating client
+   * @param {Object} data The raw data for the template
+   */
+  constructor(client, data) {
+    super(client);
+    this._patch(data);
+  }
+
+  /**
+   * Builds or updates the template with the provided data.
+   * @param {Object} data The raw data for the template
+   * @returns {GuildTemplate}
+   * @private
+   */
+  _patch(data) {
+    /**
+     * The unique code of this template
+     * @type {string}
+     */
+    this.code = data.code;
+
+    /**
+     * The name of this template
+     * @type {string}
+     */
+    this.name = data.name;
+
+    /**
+     * The description of this template
+     * @type {?string}
+     */
+    this.description = data.description;
+
+    /**
+     * The amount of times this template has been used
+     * @type {number}
+     */
+    this.usageCount = data.usage_count;
+
+    /**
+     * The ID of the user that created this template
+     * @type {string}
+     */
+    this.creatorID = data.creator_id;
+
+    /**
+     * The user that created this template
+     * @type {User}
+     */
+    this.creator = this.client.users.add(data.creator);
+
+    /**
+     * The time of when this template was created at
+     * @type {Date}
+     */
+    this.createdAt = new Date(data.created_at);
+
+    /**
+     * The time of when this template was last synced to the guild
+     * @type {Date}
+     */
+    this.updatedAt = new Date(data.updated_at);
+
+    /**
+     * The ID of the guild that this template belongs to
+     * @type {string}
+     */
+    this.guildID = data.source_guild_id;
+
+    /**
+     * The guild that this template belongs to
+     * @type {Guild}
+     */
+    this.guild = this.client.guilds.add(data.serialized_source_guild, false);
+
+    /**
+     * Whether this template has unsynced changes
+     * @type {?boolean}
+     */
+    this.unSynced = 'is_dirty' in data ? Boolean(data.is_dirty) : null;
+
+    return this;
```

Comment:
_patch usually returns void, it would be inconsistent to change this wouldn't it?

Reply:
I don't see any real downsides of having it return the instance itself, I made this change to avoid having to have some code duplication over on `edit/sync()` (`this._patch(data); return this;`). One thing to also note, is that in [`VoiceState#_patch()`](https://github.com/discordjs/discord.js/blob/master/src/structures/VoiceState.js#L71) the `VoiceState` instance is also returned over there.